systest/tests/checkpoint_test.go

```diff
@@ -102,6 +103,43 @@ func TestCheckpoint(t *testing.T) {
 	}
 	require.Empty(t, diffs)
 
+	// remove some atx from the snapshot so miners should preserve their own
+	var snapshot types.Checkpoint
+	require.NoError(t, json.Unmarshal(checkpoints[0], &snapshot))
+	// make half of the cluster preserve their own atx and chain
+	numKept := (oldSize - cl.Bootnodes()) / 2
+	var newAtxs []types.AtxSnapshot
+	preserved := map[types.NodeID]struct{}{}
+	for _, atx := range snapshot.Data.Atxs {
+		smesher := types.BytesToNodeID(atx.PublicKey)
+		if _, ok := preserved[smesher]; ok || len(preserved) < numKept {
+			preserved[smesher] = struct{}{}
+			newAtxs = append(newAtxs, atx)
+		}
+	}
+	snapshot.Data.Atxs = newAtxs
+	newSnapshot := bytes.NewBuffer(nil)
+	require.NoError(t, json.NewEncoder(newSnapshot).Encode(&snapshot))
+
+	ip, err := cl.Bootstrapper(0).Resolve(tctx)
+	require.NoError(t, err)
+	tctx.Log.Debugw("resolved bootstrapper", "ip", ip)
+	endpoint := fmt.Sprintf("http://%s:%d", ip, 80)
+	updateUrl := fmt.Sprintf("%s/updateCheckpoint", endpoint)
+	tctx.Log.Infow("submit checkpoint data", "update url", updateUrl)
+	require.NoError(t, updateCheckpointServer(tctx, updateUrl, newSnapshot.Bytes()))
+
+	queryUrl := fmt.Sprintf("%s/checkpoint", endpoint)
+	tctx.Log.Debugw("query checkpoint data", "query url", queryUrl)
+	data, err := query(tctx, queryUrl)
+	require.NoError(t, err)
+	require.True(t, bytes.Equal(newSnapshot.Bytes(), data))
```

Comment:
This helps to see what the difference is should the assertion fail:

```suggestion
	require.Equal(t, newSnapshot.String(), string(data))
```

Also imo instead of serializing the expected data to json, it would be better to deserialize the received bytes to an object and compare those two. Arguably these two are the same: `{"field1":1,"field2":2}` and `{"field2":2,"field1":1}` but comparing them as strings will cause the assertion to fail. Same with nested objects and order of elements in arrays.

Reply:
in the systest i am trying hard to be black box (not knowing the data transpired btwn nodes.
here all i care is the data is the same as the one i just updated.

Reply:
The problem is that if that assertion fails it doesn't display way:

```
FAIL TestName
     Reason: Should be true
```

Which is especially hard to fix later should the test become flaky and a failing test doesn't say why it failed, for this I think using strings and `require.Equal` would be helpful over `require.True`.

The second (less important) issue is that both values can be different but arguably the test should still pass in my example above.

Reply:
the only reason i serialize data into json is so that i can manipulate the snapshot such that half of the miners need to preserve their own atxs after checkpoint. i'll deserialize the byte to match the content.